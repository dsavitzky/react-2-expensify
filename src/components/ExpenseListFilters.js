import React from 'react'
import { connect } from 'react-redux'
import { setTextFilter, sortBy, setDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates'

export class ExpenseListFilters extends React.Component {

    state = {
        calendarFocused: null
    }
    
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }

    onFocusChange = calendarFocused => {
        this.setState(() => ({ calendarFocused }))
    }

    onTextChange = e => { 
        this.props.setTextFilter(e.target.value)
    }
    
    onSort = e => {
        this.props.sortBy(e.target.value)
    }

    render() {
        return (
            <div>
                <input 
                    type="text"
                    value={this.props.filters.text}
                    onChange={this.onTextChange} />
                <select 
                    name="" 
                    id="" 
                    value={this.props.filters.sortBy}
                    onChange={this.onSort}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
            </div>
        )
    }
}

const mapStateToProps = state => ({ filters: state.filters })

const mapDispatchToProps = dispatch => ({
    setTextFilter: text => dispatch(setTextFilter(text)),
    sortBy: target => dispatch(sortBy(target)),
    setStartDate: startDate => dispatch(setDate('startDate', startDate)),
    setEndDate: endDate => dispatch(setDate('endDate', endDate))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)