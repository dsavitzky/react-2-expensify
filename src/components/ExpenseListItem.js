import React from 'react'
import { Link } from 'react-router-dom'

const ExpenseListItem =  ({ description, amount, createdAt, id }) => (
    <li>
        <p>Description: <Link to={`/edit/${id}`}>{description}</Link></p>
        <p>Amount: {amount}</p>
        <p>Created at: {createdAt}</p>
    </li>
)

export default ExpenseListItem