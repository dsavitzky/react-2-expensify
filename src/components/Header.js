import React from 'react'
import { NavLink } from 'react-router-dom'

const Header = () => (
    <header>
        <h1>Expensify</h1>
        <ul>
            <li>
                <NavLink to="/" activeClassName="isActive" exact={true}>Home</NavLink>
            </li>
            <li>
                <NavLink to="/create" activeClassName="isActive">Create</NavLink>
            </li>
        </ul>
    </header>
)

export default Header
