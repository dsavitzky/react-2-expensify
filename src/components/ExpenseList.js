import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses'
import selectExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral'

export const ExpenseList = props => (

    <React.Fragment>
        <h2>You have {props.expenses.length === 1 ? '1 expense' : `${props.expenses.length} expenses`} { props.expenseTotal === 0 ? '' : `totaling ${numeral(props.expenseTotal / 100).format('$0,0.00')}`}</h2>
        <ul>
            {
                props.expenses.length === 0 ? (
                    ''
                ) : (
                    props.expenses.map((expense) => (
                    <ExpenseListItem key={expense.id} {...expense} />
            ))
                )
            }
            
        </ul>
    </React.Fragment>
)

const mapStateToProps = state => {

    const selectedExpenses = selectExpenses(state.expenses, state.filters)
    return {
        expenses: selectedExpenses,
        expenseTotal: selectExpensesTotal(selectedExpenses)
    }
}

export default connect(mapStateToProps)(ExpenseList)