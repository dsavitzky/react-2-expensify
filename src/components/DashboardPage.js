import React, { Fragment } from 'react'
import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'

const DashboardPage = () => (
    <Fragment>
        <ExpenseListFilters />
        <ExpenseList />
    </Fragment>
)

export default DashboardPage