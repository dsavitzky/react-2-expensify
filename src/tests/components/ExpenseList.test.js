import React from 'react'
import { shallow } from 'enzyme'
import { ExpenseList } from '../../components/ExpenseList'
import { expenses } from '../fixtures/expenses'
import { createSerializer } from 'enzyme-to-json';
 
expect.addSnapshotSerializer(createSerializer({mode: 'deep'}));

test('should render ExpenseList with expenses', () => {

    const wrapper = shallow(<ExpenseList expenses={expenses}/>)
    expect(wrapper).toMatchSnapshot()
})

test('should render ExpenseList with empty message', () => {

    const wrapper = shallow(<ExpenseList expenses={[]}/>)
    expect(wrapper).toMatchSnapshot()
})