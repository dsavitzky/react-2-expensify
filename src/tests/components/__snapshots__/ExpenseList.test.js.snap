// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with empty message 1`] = `
<ul>
  <p>
    No expenses
  </p>
</ul>
`;

exports[`should render ExpenseList with expenses 1`] = `
<ul>
  <ExpenseListItem
    amount={195}
    createdAt={0}
    description="Gum"
    id={1}
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={500}
    createdAt={432000000}
    description="Rent"
    id={2}
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={300}
    createdAt={864000000}
    description="Gum book"
    id={3}
    key="3"
    note=""
  />
</ul>
`;
